noir-projects/noir-contracts/contracts/amm_contract/src/main.nr

```diff
@@ -0,0 +1,371 @@
+
+        let reserve0_with_amount0_desired = token0.balance_of_public(context.this_address()).view(&mut context) as u64; // TODO: Avoid the need for these casts.
+        let reserve1_with_amount1_desired = token1.balance_of_public(context.this_address()).view(&mut context) as u64;
+
+        let reserve0 = reserve0_with_amount0_desired - amount0_desired;
+        let reserve1 = reserve1_with_amount1_desired - amount1_desired;
+
+        // Calculate the amounts to be added to the pool
+        let mut amount0 = amount0_desired;
+        let mut amount1 = amount1_desired;
+        if ((reserve0 != 0) | (reserve1 != 0)) {
+            // First calculate the optimal amount of token1 based on the desired amount of token0.
+            let amount1_optimal = get_quote(amount0_desired, reserve0, reserve1);
+            if (amount1_optimal <= amount1_desired) {
+                // Revert if the optimal amount of token1 is less than the desired amount of token1.
+                assert(amount1_optimal >= amount1_min, "INSUFFICIENT_1_AMOUNT");
+                amount0 = amount0_desired;
+                amount1 = amount1_optimal;
+            } else {
+                // We got more amount of token1 than desired so we try repeating the process but this time by quoting
+                // based on token1.
+                let amount0_optimal = get_quote(amount1_desired, reserve1, reserve0);
+                assert(amount0_optimal <= amount0_desired);
+                assert(amount0_optimal >= amount0_min, "INSUFFICIENT_0_AMOUNT");
+                amount0 = amount0_optimal;
+                amount1 = amount1_desired;
+            }
+        }
+
+        let refund_amount_token0 = amount0_desired - amount0;
+        let refund_amount_token1 = amount1_desired - amount1;
+
+        // The refund does not need to be finalized if the refund amount is 0 --> the partial note will either be wiped
+        // out from transient storage at the end of the tx (which is fine) or it will stay in public storage (which is
+        // also fine).
+        if (refund_amount_token0 > 0) {
+            token0.finalize_transfer_to_private(refund_token0_slot_commitment, refund_amount_token0).call(&mut context);
+        }
+        if (refund_amount_token1 > 0) {
+            token1.finalize_transfer_to_private(refund_token1_slot_commitment, refund_amount_token1).call(&mut context);
+        }
+
+        // Calculate the amount of liquidity tokens to mint
+        let total_supply = liquidity_token.total_supply().view(&mut context) as u64;
+        let mut liquidity: u64 = 0;
+        if (total_supply == 0) {
+            // TODO: This is using Tonelli-Shanks to compute sqrt but Uni is using babylonian method. Is it fine to use a different one?
+            // TODO: avoid the casts here. Shall we use a method natively working with some integer type?
+            liquidity = std::ec::sqrt((amount0 * amount1 - MINIMUM_LIQUIDITY) as Field) as u64;
```

Comment:
We actually don't even need to use `sqrt` here - this is only required if you want to pay protocol fees on add/remove liquidity events (which we won't, for simplicity). We can make the initial mint be whatever amount we wish.

Reply:
Cool. That simplifies it. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/8644/commits/417f07422f2d941bc98dab4d4a6b4c521ab1c039